File: Leetcode/Two Pointers/905. Sort Array By Parity.cpp
class Solution {
public:
    vector<int> sortArrayByParity(vector<int>& nums) {
//         vector<int>v;
//         // sort(nums.begin(),nums.end());
//         for(int i =0;i<nums.size();i++)
//         {
//             if(nums[i]%2==0)
//             {
//                 v.push_back(nums[i]);
//             }
            
//         }
//         for(int i=0;i<nums.size();i++)
//         {
//             if(nums[i]%2==1)
//             {
//                 v.push_back(nums[i]);
//             }
//         }
//another approach
int i=0;
int j=nums.size()-1;
int even=0;//for storing index of odd element  
int odd=nums.size()-1;
while(i<=j)
{
        if(nums[i]%2==0)
        {
           i++;
        }
        else
        {
           even=i;
           //cout<<"inside else of even part"<<even<<endl;
        }
       
        if(nums[j]%2!=0)
        {
          j--;
        }
        else
        {
            odd=j;
            // cout<<"inside else of odd part"<<odd<<endl;
        }

        
    
//checking the condition using two pointers while if even no are already in their place and that means odd are also so we can't swap that values
        if(even==i && odd==j)
        {
            int temp=nums[even];
            nums[even]=nums[odd];
            nums[odd]=temp;
       }
}
return nums;
    // return v;   
    }
};

File: Leetcode/Two Pointers/905. Sort Array By Parity.py
class Solution:
    def sortArrayByParity(self, nums: List[int]) -> List[int]:
        #faster approach
        even=list()
        odd=list()
        for x in nums:
            if x%2!=0:
                odd.append(x)
            else:
                even.append(x)
        return even+odd
            
        
        #not optimized 
        # temp=[]
        # for x in nums:
        #     if x%2==0:
        #         temp.append(x)
        # for y in nums:
        #     if y%2!=0:
        #         temp.append(y)
        # return temp
        
        
        #little bit optimized
        # return [i for i in nums if i%2==0] + [i for i in nums if i%2!=0]
        
